akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala

```diff
@@ -10,18 +10,21 @@ import akka.stream.{ FlowShape, Materializer }
 import akka.stream.stage.GraphStage
 import akka.util.ByteString
 import headers.HttpEncoding
-import akka.stream.scaladsl.{ Sink, Source, Flow }
+import akka.stream.scaladsl.{ Flow, Sink, Source }
 
 import scala.concurrent.Future
 
 trait Decoder {
   def encoding: HttpEncoding
 
-  def decode[T <: HttpMessage](message: T)(implicit mapper: DataMapper[T]): T#Self =
+  def decodeMessage(message: HttpMessage): HttpMessage#Self =
```

Comment:
I think this can still be improved. We need to fix the `Self` type in `HttpMessage` to make it work (props to @rkuhn who figured out how to tie the recursive self-type knot for `FlowOps` a while ago).

If we change the `Self` type in `HttpMessage` to this:

```scala
type Self <: HttpMessage { type Self = HttpMessage.this.Self }
```

the type signature for `decodeMessage` can become

```scala
def decodeMessage(message: HttpMessage): message.Self
```

which means that whatever `HttpMessage` type is passed in will also come out.